### PR TITLE
{AKS} Refactor aks nodepool add - part 3

### PR DESCRIPTION
**Pull request by FumingZhang**
**Description**

Refactor aks nodepool add - part 3.

- Add cluster auto scaler settings.
- Add os disk settings.

Unit Test Coverage:

> "covered_lines": 154,
> "num_statements": 160,
> "percent_covered": 96.25,
> "percent_covered_display": "96",
> "missing_lines": 6,
> "excluded_lines": 0

**Testing Guide**

**History Notes**

[Component Name 1] BREAKING CHANGE: `az command a`: Make some customer-facing breaking change
[Component Name 2] `az command b`: Add some customer-facing feature

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [ ] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
AKS

File: src/azure-cli/azure/cli/command_modules/acs/agentpool_decorator.py [hunk 1]
```diff
@@ -200,6 +242,94 @@ def get_max_surge(self):
...
+            min_count,
+            max_count,
+            decorator_mode=DecoratorMode.CREATE,
+        )
+        return node_count, enable_cluster_autoscaler, min_count, max_count
+
+    def get_node_osdisk_size(self) -> Union[int, None]:
+        """Obtain the value of node_osdisk_size.
+
+        Note: SDK performs the following validation {'maximum': 2048, 'minimum': 0}.
```
Comment (andyliuliming) [inline src/azure-cli/azure/cli/command_modules/acs/agentpool_decorator.py:299]: link to the SDK?

File: src/azure-cli/azure/cli/command_modules/acs/agentpool_decorator.py [hunk 2]
```diff
@@ -200,6 +242,94 @@ def get_max_surge(self):
...
+
+        This function will normalize the parameter by default. The parameter will be converted to int.
+
+        :return: int or None
+        """
+        # read the original value passed by the command
+        node_osdisk_size = self.raw_param.get("node_osdisk_size")
```
Comment (andyliuliming) [inline src/azure-cli/azure/cli/command_modules/acs/agentpool_decorator.py:306]: is there any place to check the format of the string?

File: src/azure-cli/azure/cli/command_modules/acs/agentpool_decorator.py [hunk 3]
```diff
@@ -200,6 +242,94 @@ def get_max_surge(self):
...
+
+    def get_node_osdisk_type(self) -> Union[str, None]:
+        """Obtain the value of node_osdisk_type.
+
+        :return: string or None
+        """
+        # read the original value passed by the command
+        node_osdisk_type = self.raw_param.get("node_osdisk_type")
+        # try to read the property value corresponding to the parameter from the `agentpool` object
+        if self.agentpool and self.agentpool.os_disk_type is not None:
```
Comment (andyliuliming) [inline src/azure-cli/azure/cli/command_modules/acs/agentpool_decorator.py:326]: then we can not update the node_osdisk_type anymore?

**Review by andyliuliming — CHANGES_REQUESTED** [review]

File: src/azure-cli/azure/cli/command_modules/acs/agentpool_decorator.py (hunk 2 above)
Comment (FumingZhang) [reply src/azure-cli/azure/cli/command_modules/acs/agentpool_decorator.py:306]: Fixed in commit [#b9cf823](https://github.com/Azure/azure-cli/pull/21598/commits/b9cf8238fac98d5bb51e3a0f6b74a4fca8247438), added type restrictions to parameter declarations, cli will properly handle type errors.

Also fixed incorrect parameter declaration for `node_osdisk_type`.

File: src/azure-cli/azure/cli/command_modules/acs/agentpool_decorator.py (hunk 1 above)
Comment (FumingZhang) [reply src/azure-cli/azure/cli/command_modules/acs/agentpool_decorator.py:299]: Emm, this note is just a reminder, will it be too heavy to add a link?

**Review by andyliuliming — APPROVED** [review]

**Review by zhoxing-ms — APPROVED** [review]

**Pull request merged**